sklearn/utils/arrayfuncs.pyx

```diff
@@ -4,20 +4,16 @@ Small collection of auxiliary functions that operate on arrays
 """
 
 cimport numpy as cnp
-import  numpy as np
+import numpy as np
+from numpy cimport ndarray
 from cython cimport floating
 from libc.math cimport fabs
 from libc.float cimport DBL_MAX, FLT_MAX
 
 from ._cython_blas cimport _copy, _rotg, _rot
 
-ctypedef cnp.float64_t DOUBLE
 
-
-cnp.import_array()
-
-
-def min_pos(cnp.ndarray X):
+def min_pos(floating[:] X):
```

Comment:
Hey @rprkh, thank you for this PR! I think this one is tricky. What I suggest is:
```python
def min_pos(const floating[:] X):
    """Find the minimum value of an array over positive values

    Returns the maximum representable value of the input dtype if none of the
    values are positive.
    """
    if np.asarray(X).dtype == np.float32:
        return _min_pos[float](<float *> &X[0], X.size)
    elif np.asarray(X).dtype == np.float64:
        return _min_pos[double](<double *> &X[0], X.size)
    else:
        raise ValueError('Unsupported dtype for array X')
```
This should solve your Cython compiling error. Note that const-qualified fuse type memoryviews are now supported thanks to #25342 :)

Reply:
Perfect explanation! :)

If I were to add something, it would be indicating that `const`-qualification (the use of `const`) allows supporting read-only data correctly (I do not think we need to explain the recent support of `const`-qualified fused type memoryviews).